src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,280 @@
+};
+
+export const getDefaultManager = once(() => {
+	return new DeviceOnlineStateManager();
+});
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class NoMessagesToConsumeSignal extends Error {}
+
+class DeviceOnlineStateManager extends EventEmitter {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		super();
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: process.env.REDIS_HOST,
+			port: parseInt(process.env.REDIS_PORT || '6379'),
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueue({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch((err: Error) => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				this.consume();
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Online);
+	}
+
+	private deviceOffline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Offline);
+	}
+
+	private deviceTimeout(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Timeout);
+	}
+
+	private consume() {
+		let delayBeforeNextConsume = 1;
+
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessage({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: 30, // prevent other consumers seeing the same message (if any) for 30 secs (prevents multiple API agents from processing it)...
+			})
+			.tap((msg: any) => {
+				// no message! this is thrown to restart the consume process...
+				if (!msg || !msg.id) {
+					delayBeforeNextConsume = 1000;
+					throw new NoMessagesToConsumeSignal();
+				}
+			})
+			.then((msg: any) => {
+				const { id, message } = msg;
+
+				return Bluebird.try(() => {
+					const { uuid, nextState } = JSON.parse(message);
+
+					// raise and event for the state change...
+					switch (nextState) {
+						case DeviceOnlineStates.Timeout:
+							this.scheduleChangeOfStateForDevice(
+								uuid,
+								DeviceOnlineStates.Timeout,
+								DeviceOnlineStates.Offline,
+								15, // put the device into a timeout state for 15 sec if it misses it's scheduled heartbeat window... then mark as offline
+							);
+							this.deviceTimeout(uuid);
+							break;
+						case DeviceOnlineStates.Offline:
+							this.deviceOffline(uuid);
+							break;
+						default:
+							// nothing should be raised here...
+							break;
+					}
+				}).then(() =>
+					this.rsmq.deleteMessage({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id: id,
+					}),
+				);
+			})
+			.catch(NoMessagesToConsumeSignal, noop)
```

Comment:
It would be better to wrap this into the `.catch` below because there's a cost to this sugar which could add up since I believe this function is called a lot

Reply:
Done